kernel/arch/dreamcast/hardware/maple/keyboard.c

```diff
@@ -23,16 +26,88 @@ repeat handling.
 
 */
 
+/*  Size of a keyboard queue.
+
+    Each keyboard queue will hold this many elements. Once the queue fills, no
+    new elements will be placed on the queue. As long as you check the queue
+    relatively frequently, the default of 16 should be plenty.
+
+    This MUST be a power of two.
+*/
+#define KBD_QUEUE_SIZE 16
+
+/* Keyboard raw condition structure.
+
+    This structure is what the keyboard responds with as its current status.
+*/
+typedef struct kbd_cond {
+    kbd_mods_t modifiers;    /**< \brief Bitmask of set modifiers. */
+    kbd_leds_t leds;         /**< \brief Bitmask of set LEDs. */
+    kbd_key_t  keys[KBD_MAX_PRESSED_KEYS];      /**< \brief Key codes for currently pressed keys. */
+} kbd_cond_t;
+
+typedef struct kbd_state_private {
+    kbd_state_t base;
+
+    /* Repeated key press state data. */
+    struct {
+        kbd_key_t key;    /* Last key held which will repeat. */
+        uint64_t timeout; /* Time the next repeat will trigger. */
+    } repeater;
+
+    uint32_t key_queue[KBD_QUEUE_SIZE];
+    size_t queue_tail;          /* Key queue tail. */
+    size_t queue_head;          /* Key queue head. */
+    volatile size_t queue_len;  /* Current length of queue. */
+} kbd_state_private_t;
+
+static struct {
+    kbd_event_handler_t cb;
+    void               *ud;
+} event_handler = {
+    NULL, NULL
+};
+
+void kbd_set_event_handler(kbd_event_handler_t callback, void *user_data) {
+    event_handler.cb = callback;
+    event_handler.ud = user_data;
+}
+
 /* These are global timings for key repeat. It would be possible to put
     them in the state, but I don't see a reason to.
     It seems unreasonable that one might want different repeat
     timings set on each keyboard.
     The values are arbitrary based off a survey of common values. */
-uint16 kbd_repeat_start = 600, kbd_repeat_interval = 20;
+static struct {
+    uint16_t start;
+    uint16_t interval;
+} repeat_timing = {
+    600, 20
+};
```

Comment:
Yeah, I really didn't see any reason to duplicate them over each keyboard state or convolute the API with making the user configure them all for each individual keyboard. It seems janky that an application would want different timeout and button repeat mechanics for each keyboard device, and it's saving RAM (well, it will in the future when maple devices are dynamically allocated)... Still felt like the way to go to me.